### ISSUE TITLE
arocc doesn't follow subtle rules in preprocessing expansion

### ISSUE THREAD

**Issue by doppioandante** [Vexu/arocc#30]
Hello, I'm opening this issue mostly to clear up my understanding of these issues. Consider
```c
#define f(a) a
#define z z[0]

f(f(z))
```
with `gcc -E` this expands to `z[0]`. With #29 , this expands to `z[0][0][0]`.
From the current C standard draft, 6.10.3.4 (Rescanning and further replacement)

> If the name of the macro being replaced is found during this scan of the replacement list (not
> including the rest of the source file’s preprocessing tokens), it is not replaced. Furthermore, if any
> nested replacements encounter the name of the macro being replaced, it is not replaced. *These
> nonreplaced macro name preprocessing tokens are no longer available for further replacement even
> if they are later (re)examined in contexts in which that macro name preprocessing token would
> otherwise have been replaced.*

I guess the last sentence is what makes the difference in this case.
I would like to try to fix this myself.

**Comment by doppioandante** [comment]
Ok, I've explored this issue a bit. Considering how long ago the C preprocessor was invented, I thought that the convoluted wording in the standard was describing a simple and serial algorithm with some quirks.
Armed  with pen and paper I tried to manually expand some tricky test cases that I devised, but unfortunately I found out that in order to avoid recursion the preprocessor should keep track of what macros have been expandend when recursing, in order to not expand those pieces of the source again if the same macro name is met in that context.

That, or gcc is bugged, which I think is less likely.
I'll try implementing the algorithm in the next few days and make a PR. Solving this and #31 should be done in the same pass.
